File: chapt5/chapt5_ws/src/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformListener, Buffer
from tf_transformations import euler_from_quaternion
class TFListener(Node):
    def __init__(self, node_name):
        super().__init__(node_name)
        self.buffer_ = Buffer()
        self.tf_listener_ = TransformListener(self.buffer_, self)
        self.timer_ = self.create_timer(1, self.get_tf)
    def get_tf(self):
        try:
            target_frame = 'laser_link'
            source_frame = 'odom'
            result = self.buffer_.lookup_transform(target_frame, source_frame, rclpy.time.Time(seconds=0.0), rclpy.time.Duration(seconds=1.0))
            transform = result.transform
            self.get_logger().info(f'Get tf {source_frame} in {target_frame}: {transform.translation}')
            self.get_logger().info(f'Get rot: {transform.rotation}')
            rpy = euler_from_quaternion([transform.rotation.x, transform.rotation.y, transform.rotation.z, transform.rotation.w])
            self.get_logger().info(f'Get rot(RPY): {rpy}')
        except Exception as e:
            self.get_logger().warn(f'Could not get tf, reason: {e}')
            


def main():
    rclpy.init()
    node = TFListener('tf_listener')
    rclpy.spin(node)
    rclpy.shutdown()

File: chapt5/chapt5_ws/src/demo_cpp_tf/src/tf_listener.cpp
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_ros/transform_listener.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <chrono>
#include <tf2/utils.h>
using namespace std::chrono_literals;
using TFL = tf2_ros::TransformListener;
class TFListener : public rclcpp::Node
{
public:
    TFListener(const std::string &node_name) : Node(node_name)
    {
        timer_ = this->create_wall_timer(1000ms, std::bind(&TFListener::get_tf, this));
        tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<TFL>(*tf_buffer_);
    }
    void get_tf()
    {
        std::string target_frame = "map";
        std::string source_frame = "laser_link";
        try
        {
            const geometry_msgs::msg::TransformStamped tfs = tf_buffer_->lookupTransform(
                target_frame, source_frame, this->get_clock()->now(), rclcpp::Duration::from_seconds(1.0)
            );
            RCLCPP_INFO(this->get_logger(), "Got transform from %s in %s: translation(%.2f, %.2f, %.2f), rotation(%.2f, %.2f, %.2f, %.2f)",
                source_frame.c_str(), target_frame.c_str(), 
                tfs.transform.translation.x,
                tfs.transform.translation.y,
                tfs.transform.translation.z,
                tfs.transform.rotation.x,
                tfs.transform.rotation.y,
                tfs.transform.rotation.z,
                tfs.transform.rotation.w);
            double roll, pitch, yaw;
            tf2::getEulerYPR(tfs.transform.rotation, yaw, pitch, roll);
            RCLCPP_INFO(this->get_logger(), "Euler angles: roll: %.2f, pitch: %.2f, yaw: %.2f", roll, pitch, yaw);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Could not get transform: %s", e.what());
        }
    }
private:
    std::shared_ptr<TFL> tf_listener_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;

};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    std::shared_ptr<TFListener> node = std::make_shared<TFListener>("tf_listener");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}